spec/codebreaker_spec.rb

```diff
@@ -1,9 +1,153 @@
+    end
+
+    it 'take one number as hint and remove it from array' do
+      expect { subject.send(:take_a_hint!) }.to change { subject.instance_variable_get(:@hints).size }.by(-1)
+    end
+
+    it 'decrement hints count by one' do
+      expect { subject.send(:take_a_hint!) }.to change { subject.instance_variable_get(:@hints_count) }.by(-1)
+    end
+  end
+
+  context '#hint' do
+    it 'tells about missing hints' do
+      subject.instance_variable_set(:@hints_count, 0)
+      expect { subject.send(:hint) }.to output("Hints are over\n").to_stdout
+    end
+
+    it 'does one hint' do
+      subject.instance_variable_set(:@hints, [1, 2, 3, 4])
+      subject.instance_variable_set(:@hints_count, 2)
+      expect { subject.send(:hint) }.to output("4\n").to_stdout
+    end
   end
 
-  it "does something useful" do
-    expect(false).to eq(true)
+  context '#handle_answer' do
+    it 'calls a hint' do
+      subject.instance_variable_set(:@user_answer, 'hint')
+      allow(subject).to receive(:hint)
+      expect(subject).to receive(:hint)
+      subject.send(:handle_answer)
+    end
+
+    it 'calls validate method' do
+      subject.instance_variable_set(:@user_answer, 'test')
+      expect(subject).to receive(:validate_answer)
+      subject.send(:handle_answer)
+    end
+
+    it 'handles correct number' do
+      subject.instance_variable_set(:@user_answer, '1234')
+      expect(subject).to receive(:handle_code)
+      subject.send(:handle_answer)
+    end
+  end
+
+  context '#validate_answer' do
+    it 'tells about input error' do
+      expect { subject.send(:validate_answer) }.to output("Put 4-digital number\n").to_stdout
+    end
+
+    it 'returns true' do
+      subject.instance_variable_set(:@user_answer, '1234')
+      expect(subject.send(:validate_answer)).to be true
+    end
+  end
+
+  context '#setup_the_difficulty' do
+    it 'setup easy level of difficulty' do
+      subject.instance_variable_set(:@difficulty, Game::DIFFICULTY[:easy])
+      subject.send(:setup_the_difficulty)
+      expect(subject.instance_variable_get(:@attempts_count)).to eq(Game::DIFFICULTY.dig(:easy, :attempts))
+      expect(subject.instance_variable_get(:@hints_count)).to eq(Game::DIFFICULTY.dig(:easy, :hints))
+    end
+
+    it 'setup medium level of difficulty' do
+      subject.instance_variable_set(:@difficulty, Game::DIFFICULTY[:medium])
+      subject.send(:setup_the_difficulty)
+      expect(subject.instance_variable_get(:@attempts_count)).to eq(15)
+      expect(subject.instance_variable_get(:@hints_count)).to eq(2)
+    end
+
+    it 'setup hard level of difficulty' do
+      subject.instance_variable_set(:@difficulty, Game::DIFFICULTY[:hard])
+      subject.send(:setup_the_difficulty)
+      expect(subject.instance_variable_get(:@attempts_count)).to eq(10)
+      expect(subject.instance_variable_get(:@hints_count)).to eq(1)
+    end
+  end
+
+  context '#difficulty' do
+    it "tells about spelling error and calls itself again" do
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.